homeassistant/components/mqtt/client.py

```diff
@@ -918,36 +937,33 @@ def _async_mqtt_on_connect(
             result_code,
         )
 
-        self.hass.async_create_task(self._async_resubscribe())
-
+        self._async_queue_resubscribe()
+        birth: dict[str, Any]
         if birth := self.conf.get(CONF_BIRTH_MESSAGE, DEFAULT_BIRTH):
-
-            async def publish_birth_message(birth_message: PublishMessage) -> None:
-                await self._ha_started.wait()  # Wait for Home Assistant to start
-                await self._discovery_cooldown()  # Wait for MQTT discovery to cool down
-                # Update subscribe cooldown period to a shorter time
-                self._subscribe_debouncer.set_timeout(SUBSCRIBE_COOLDOWN)
-                await self.async_publish(
-                    topic=birth_message.topic,
-                    payload=birth_message.payload,
-                    qos=birth_message.qos,
-                    retain=birth_message.retain,
-                )
-
             birth_message = PublishMessage(**birth)
             self.config_entry.async_create_background_task(
```

Comment:
There is a race here if we disconnect while the task is running, but thats existing and should be solved in another P

We should hold a refrence to the task and cancel it on disconnect and than await it suppressing cancellederror